cni/pkg/nodeagent/net.go

```diff
@@ -312,11 +312,17 @@ func addPodToHostNSIpset(pod *corev1.Pod, podIPs []netip.Addr, hostsideProbeSet
 	for _, pip := range podIPs {
 		// Add to host ipset
 		log.Debugf("adding pod %s probe to ipset %s with ip %s", pod.Name, hostsideProbeSet.Prefix, pip)
-		// Add IP/port combo to set. Note that we set Replace to true - a pod ip/port combo already being
-		// in the set is perfectly fine, and something we can always safely overwrite, so we will.
-		if err := hostsideProbeSet.AddIP(pip, ipProto, podUID, true); err != nil {
+		// Add IP/port combo to set. Note that we set Replace to false here - we _did_ previously
+		// set it to true, but in theory that could mask weird scenarios where K8S triggers events out of order ->
+		// an add(sameIPreused) then delete(originalIP).
+		// Which will result in the new pod starting to fail healthchecks.
+		//
+		// Since we purge on restart of CNI, and remove pod IPs from the set on every pod removal/deletion,
+		// we _shouldn't_ get any overwrite/overlap, unless something is wrong and we are asked to add
+		// a pod by an IP we already have in the set (which will give an error, which we want).
+		if err := hostsideProbeSet.AddIP(pip, ipProto, podUID, false); err != nil {
```

Comment:
So if we hit the error, we fail to the CNI layer which retries and now the ordering should be correct. Right? It will recover?

Reply:
Right. Inelegant, but it will work itself out (if the hypothesis about the cause is in fact correct)